tornado/http1connection.py

```diff
@@ -180,7 +183,7 @@ def _read_message(self, delegate):
             self._disconnect_on_finish = not self._can_keep_alive(
                 start_line, headers)
             need_delegate_close = True
-            with _ExceptionLoggingContext(app_log):
+            with _ExceptionLoggingContext(app_log, pass_exception=httputil._DelegateNotFoundException):
                 header_future = delegate.headers_received(start_line, headers)
```

Comment:
I think we could put a new `try/except _DelegateNotFoundException` block around the call to `headers_received` (including `yield header_future`), moving the 404 generation here, and then we wouldn't need to modify ExceptionLoggingContext.